src/Libraries/Microsoft.Extensions.AI.Evaluation.NLP/Common/BLEUAlgorithm.cs

```diff
@@ -0,0 +1,197 @@
+    }
+
+    internal static double BrevityPenalty(int closestRefLength, int hypLength)
+    {
+        if (hypLength <= 0)
+        {
+            return 0.0;
+        }
+
+        if (closestRefLength <= 0 || hypLength > closestRefLength)
+        {
+            return 1.0;
+        }
+
+        return Math.Exp(1 - ((double)closestRefLength / hypLength));
+    }
+
+    internal static RationalNumber ModifiedPrecision(IEnumerable<IEnumerable<string>> references, IEnumerable<string> hypothesis, int n = 1)
+    {
+        if (n <= 0)
+        {
+            Throw.ArgumentOutOfRangeException(nameof(n), $"`{nameof(n)}` must be greater than zero.");
+        }
+
+        if (!references.Any() || !hypothesis.Any())
+        {
+            return new RationalNumber(0, 0);
+        }
+
+        var hyp = hypothesis.CreateNGrams(n);
+        var hypCounts = new MatchCounter<NGram<string>>(hyp);
+
+        Dictionary<NGram<string>, int> maxCounts = [];
+
+        foreach (var rf in references)
+        {
+            IEnumerable<NGram<string>> refGrams = rf.CreateNGrams(n);
+            var refCounts = new MatchCounter<NGram<string>>(refGrams);
+
+            foreach (var ct in refCounts)
+            {
+                if (maxCounts.TryGetValue(ct.Key, out int val))
+                {
+                    maxCounts[ct.Key] = Math.Max(val, ct.Value);
+                }
+                else
+                {
+                    maxCounts[ct.Key] = ct.Value;
+                }
+            }
+        }
+
+        Dictionary<NGram<string>, int> clippedCounts = [];
+        foreach (var h in hypCounts)
+        {
+            if (maxCounts.TryGetValue(h.Key, out var v))
+            {
+                clippedCounts[h.Key] = Math.Min(h.Value, v);
+            }
+            else
+            {
+                // If the hypothesis n-gram is not in any reference, it is clipped to 0.
+                clippedCounts[h.Key] = 0;
+            }
+        }
+
+        int numerator = clippedCounts.Values.Sum();
+        int denominator = Math.Max(1, hypCounts.Sum());
+
+        return new RationalNumber(numerator, denominator);
+    }
+
+    /// <summary>
+    /// Generate an n-sized array of equal weights that sum to 1.0.
+    /// </summary>
+    /// <param name="n">Number of weights to return.</param>
+    /// <returns>Array of equal sized values that sum to 1.0.</returns>
+    internal static double[] EqualWeights(int n)
+    {
+        if (n <= 0)
+        {
+            Throw.ArgumentOutOfRangeException(nameof(n), $"'{nameof(n)}' must be greater than zero.");
+        }
+
+        double[] weights = new double[n];
+        for (int i = 0; i < n; i++)
+        {
+            weights[i] = 1.0 / n;
+        }
+
+        return weights;
+    }
+
+    internal static readonly double[] DefaultBLEUWeights = EqualWeights(4);
+
+    internal static double SentenceBLEU(IEnumerable<IEnumerable<string>> references, IEnumerable<string> hypothesis,
```

Comment:
Singular -> plural 

```suggestion
    internal static double SentenceBLEU(IEnumerable<IEnumerable<string>> references, IEnumerable<string> hypotheses,
```

Reply:
Ah never mind - just realized that this is the set of words in the hypothesis as opposed to multiple hypotheses :)